### PR TITLE
2015-03-15 ES Hotspots

### PR DESCRIPTION
**Pull request by aewhite**
@kennycason @addamh review please 

You will probably want to review this in an HTML preview.

### CODE REVIEW

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+This post will be exploring the statistical nature of log normal distributed data with an emphasis on how it affects the
+ design of Elasticsearch clusters. However, the data and conclusions should be relevant beyond the scope of Elasticsearch.
+
+<!-- loading MathJax here unless there is a better way elsewhere -->
+<script type="text/javascript"
+  src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
+</script>
+
+First, let's define some terms. At Datarank, we track "topics". A topic is nothing more than a collections of comments
+that match some criteria. These criteria can be rather broad to very specific. A shard is a slice of an Elasitcsearch
```
Comment (sawyercade) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: Elasticsearch is misspelled.

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+ $$P(T)=\frac{1}{x\sigma\sqrt{2\pi}}e^{-\frac{(\ln x - \mu)^2}{2 \sigma^2}};\;\mu=9.022\,\mathrm{and}\,\sigma=2.932$$
+
+As an interesting aside, this is the same equation used to model the spread of highly communicable epidemics. Since
+topics track abstract concepts, this may provide some good evidence for modeling highly shared content, such as ads and
+memes, as viruses.
+
+## Shard Sizes
+
+For performance reasons, we like to map a topic to a minimum number of shards, ideally one. To that end we use
```
Comment (sawyercade) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: You might want to expound a little bit here on why specifically we like to map a topic to a minimum number of shards. Is it for faster indexing or lookup?

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+Elasticsearch's routing feature and use the topic's numerical id as the routing key. It's important to note that routing
+is static so that document placement is predictable and stable as more data is added to the index.
+
+Since a shard is simply a collections of topics, it's size PDF can be expressed as follows.
+
+ $$P(S) = T + T + T + ... + T$$
+
+Unfortunately, there is no closed form for the sum of the log-normal distribution. However, we can calculate the expected
+  value for $$P(S)$$ by noting that topic sizes are independent. Therefore we can say that
```
Comment (sawyercade) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: For some reason, when viewed in an HTML preview, there's no space between P(S) and "noting".

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown (hunk 1 above)
Comment (aewhite) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: Noted

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+we can see that the variance increases as well; however, this rate is not linear.
+
+## Index Balance
+
+Since an index is simply a collection of shards we can start to talk about balancing an index across many shards. Ideally,
+shards should map one to one to physical nodes. However, this becomes impractical with log-normally distributed data as
+we will soon discover.
+
+First, let's define how balanced an index is by the ratio of the smallest shard to the largest shard. A ratio of 1 would
+be perfect; whereas, a ration of 2 would mean that the largest shard is 2x larger than the smallest shard. We use this
```
Comment (sawyercade) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: ~~ration~~ ratio

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown (hunk 2 above)
Comment (aewhite) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: I think that would distract from the core of the article. "We do it for performance reasons" seems like a reasonable assumption to make and justifying it seems unnecessary since it is a semi-known best practice that you want the minimum number of shards involved in any query.

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown (hunk 3 above)
Comment (aewhite) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: This seems to be a known bug for certain browsers that I can't seem to find a workaround for. I'm going to call this good for now since it's pretty minor.

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown (hunk 4 above)
Comment (aewhite) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: Noted

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+ratio over something like a standard deviation due to importance of overloaded nodes. In general 1.25 - 1.50 are quite
+good.
```
Comment (sawyercade) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: ~~1.25 - 1.50 are quite good.~~

a ratio of 1.25 - 1.50 is quite well balanced.

Comment (aewhite) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: Agreed

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+Unfortunately, there is no closed form for the sum of the log-normal distribution. However, we can calculate the expected
+  value for $$P(S)$$ by noting that topic sizes are independent. Therefore we can say that
+  $$\mu_S = numberOfTopics*\mu_T = numberOfTopics*e^{\mu+\sigma^2/2}$$
+
+The graph below illustrates PDFs for shard sizes of a variety of topic groupings. Note, for this graph we used random
+variant sampling from $$P(T)$$ and then a smooth kernel distribution to build the histograms.
+
+<img src="/images/posts/2015-06-16/pdf-single-shard-size.png">
+
+From the graph, it is clear that the mode (and the mean) increase linearly as the number of topics increases. In addition,
```
Comment (sawyercade) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: Could you change the x-axis label to Shard Size, for clarity?

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+## Shard Sizes
+
+For performance reasons, we like to map a topic to a minimum number of shards, ideally one. To that end we use
+Elasticsearch's routing feature and use the topic's numerical id as the routing key. It's important to note that routing
+is static so that document placement is predictable and stable as more data is added to the index.
+
+Since a shard is simply a collections of topics, it's size PDF can be expressed as follows.
+
+ $$P(S) = T + T + T + ... + T$$
```
Comment (sawyercade) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: Shouldn't this be P(S) = P(T) + P(T) + ... + P(T)?

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown (hunk 6 above)
Comment (aewhite) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: Yes.

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown (hunk 7 above)
Comment (aewhite) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: I've seen it written a couple of ways and probably S = T + T + T is better (see http://leo.ugr.es/pgm2012/submissions/pgm2012_submission_6.pdf)

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+changes with various topic and shard count configurations.
+
+<img src="/images/posts/2015-06-16/pdf-10-shard-ratio.png">
+
+<img src="/images/posts/2015-06-16/pdf-shard-ratio-10k.png">
+
+First, note that the imbalance is least when we have fewer denser shards. This implies that horizontal scaling will
+actually make the problem worse. Second, as more topics are added to an existing cluster, it should naturally become
+more balanced over time.
+
```
Comment (sawyercade) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: This whole paragraph is the exact opposite of what I would intuitively expect! Unless all of the topics were assumed to be the same size, I would think that splitting the distributing topics across more shards would result in the mode of the ratio PDF shifting left, not right, although I would assume that the variance would be greater as is the case in this graph. Did you use random variant sampling from P(T) here as well?

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+Since the cluster naturally becomes more balanced with volume the problem is solved, right? The problem manifests itself
+when two critical aspects are noted. First, even in optimistic cases the ratio is still quite high. These high ratios
+mean that the cluster runs the risk of overloading a node during a rebalancing. Second, very large shards can run into
+performance issues for aggregation of very large topics.
+
+Luckily there are some workarounds to help mitigate the problem. First, separate out large topics; then use a resource
```
Comment (sawyercade) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: ~~separate out large topics~~

separate large topics into their own single-topic indexes

Comment (aewhite) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: agreed

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+
+<img src="/images/posts/2015-06-16/pdf-large-topic-excludes.png">
+
+At σ = 3 almost nothing is being excluded and the mode imbalance ratio is around 2.75. Exclude just 2% of the topics
+(σ = 2.25) and the mode imbalance ratio drops to 1.8; however, with thousands of topics, this produces several dozen
+separate indices that have to be managed. At Datarank we use a σ of 2.35 and get fair results out of the default allocator.
+
+## Conclusion
+
+In this post we have shown that hotspots in statically routed documents that fit into log-normally distrusted topic sizes
```
Comment (sawyercade) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: ~~distrusted~~

distributed

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown (hunk 8 above)
Comment (aewhite) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: Yes, this used random varient sampling as well. And yes, this does seem odd. The best explanation I can give is that with more shards you give large topics more of a chance to cluster together (birthday paradox again). And even small clusters of large topics cause huge imbalances.

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown (hunk 10 above)
Comment (aewhite) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: agreed.

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+will always produce hotspots. However, by separating out very large topics, the imbalance of shard sizes decreases
+dramatically. However, this is still far from ideal due to how the default Elasticsearch allocator works. This is why
+Datarank has developed a custom resource aware allocator that will be released and commented on soon.
```
Comment (sawyercade) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: ~~resource aware~~

resource-aware

~~on~~

upon

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+
+At σ = 3 almost nothing is being excluded and the mode imbalance ratio is around 2.75. Exclude just 2% of the topics
+(σ = 2.25) and the mode imbalance ratio drops to 1.8; however, with thousands of topics, this produces several dozen
+separate indices that have to be managed. At Datarank we use a σ of 2.35 and get fair results out of the default allocator.
+
+## Conclusion
+
+In this post we have shown that hotspots in statically routed documents that fit into log-normally distrusted topic sizes
+will always produce hotspots. However, by separating out very large topics, the imbalance of shard sizes decreases
+dramatically. However, this is still far from ideal due to how the default Elasticsearch allocator works. This is why
```
Comment (sawyercade) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: ~~how the default Elasticsearch allocator works~~

the default Elasticsearch allocator being resource-blind.

**Comment by sawyercade** [comment]
:+1:

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown (hunk 2 above)
Comment (kennycason) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: I'd say worst-case you could add footnotes.

// el blog post. blah blah blah blah [1]. mas data del blog post
[1] explanation/site for blah blah blah blah

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+
+This post will be exploring the statistical nature of log normal distributed data with an emphasis on how it affects the
+ design of Elasticsearch clusters. However, the data and conclusions should be relevant beyond the scope of Elasticsearch.
+
+<!-- loading MathJax here unless there is a better way elsewhere -->
+<script type="text/javascript"
+  src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
+</script>
+
+First, let's define some terms. At Datarank, we track "topics". A topic is nothing more than a collections of comments
```
Comment (kennycason) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: collection* of comments

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+that match some criteria. These criteria can be rather broad to very specific. A shard is a slice of an Elasticsearch
```
Comment (kennycason) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: These criteria can range from being rather broad to very specific.

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+topics track abstract concepts, this may provide some good evidence for modeling highly shared content, such as ads and
+memes, as viruses.
+
+## Shard Sizes
+
+For performance reasons, we like to map a topic to a minimum number of shards, ideally one. To that end we use
+Elasticsearch's routing feature and use the topic's numerical id as the routing key. It's important to note that routing
+is static so that document placement is predictable and stable as more data is added to the index.
+
+Since a shard is simply a collections of topics, it's size PDF can be expressed as follows.
```
Comment (kennycason) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: collection*

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown (hunk 6 above)
Comment (kennycason) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: units for shard size?

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+Since an index is simply a collection of shards we can start to talk about balancing an index across many shards. Ideally,
+shards should map one to one to physical nodes. However, this becomes impractical with log-normally distributed data as
+we will soon discover.
+
+First, let's define how balanced an index is by the ratio of the smallest shard to the largest shard. A ratio of 1 would
+be perfect; whereas, a ratio of 2 would mean that the largest shard is 2x larger than the smallest shard. We use this
+ratio over something like a standard deviation due to importance of overloaded nodes. In general a ratio of 1.25 - 1.50
+is quite well balanced
+
+Side note: The default Elasticsearch shard allocator does not take shard size into account when distributing shards
```
Comment (kennycason) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: I actually don't think this is a side note, but think this is a pretty huge reason to justify why we built this in the first place.

**Comment by kennycason** [comment]
:+1: That was an awesome read :)

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown (hunk 6 above)
Comment (aewhite) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: The unit is #of documents, I'll update it.

File: _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown (hunk 16 above)
Comment (aewhite) [inline _posts/2015-06-16-analysis-of-hotspots-in-clusters-of-log-normally-distributed-data.markdown]: I think it is critical to the problem that @sawyercade addresses but not to the existence of hotspots per say. Let me do some re-wording.

**Comment by aewhite** [comment]
@kennycason @sawyercade More PR changes made. @addamh if you approve feel free to merge and publish.
